### PR TITLE
#164798200 User should receive descriptive validation errors messages on registration

### PR DESCRIPTION
**Pull request by felixkiryowa**
#### What does this PR do?

- Enables users receive descriptive error validation messages on signing up

#### Description of Task to be completed?
-  When a user is registering for an account and somehow they don't meet the requirements required for somebody to create an account, then they should be able to get a message as to why they are not able to fulfill their request

#### How should this be manually tested?

- Fetch this branch `git fetch origin ft-receive-descriptive-validation-errors-164798200`
- Checkout to the branch `git checkout ft-receive-descriptive-validation-errors-164798200`
- Run `python manage.py runserver`
- Run register user endpoint

#### What are the relevant pivotal tracker stories?
[#164798200](https://www.pivotaltracker.com/story/show/164798200)

#### Postman Collection
[Run in Postman](https://app.getpostman.com/run-collection/cb67ec77f04e89b40156)

### CODE REVIEW

File: authors/apps/authentication/tests/test_base.py [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
 from authors.apps.authentication.tests.test_data import (
-    login_credentials_data, 
+    login_credentials_data,
     login_data
-    )
+)
+
 
-class BaseTestCase(APITestCase):
+class BaseTestCase(TestCase):
```
Comment (ja-odur) [inline authors/apps/authentication/tests/test_base.py:12]: @felixkiryowa Please undo the changes in this `test_base.py` file.

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -28,8 +31,61 @@ class Meta:
...
+        keys = self.fields.keys()
+
+        for key in values.copy():
+            values.pop(key) if key not in keys else None
+        return values
+
+
     def create(self, validated_data):
         # Use the `create_user` method we wrote earlier to create a new user.
+
```
Comment (ja-odur) [inline authors/apps/authentication/serializers.py]: @felixkiryowa please remove this unnecessary space

File: Procfile [hunk 3]
```diff
@@ -1 +1,2 @@
 web: gunicorn authors.wsgi
+release: python manage.py migrate --noinput
```
Comment (ja-odur) [inline Procfile:2]: @felixkiryowa this change is no longer necessary.

File: authors/apps/authentication/tests/test_user_registration.py [hunk 4]
```diff
@@ -31,47 +39,100 @@ def test_register_a_user_with_no_data(self):
         """
         Method to test if posted user registration user object contains  no data
         """
-        
+
```
Comment (ja-odur) [inline authors/apps/authentication/tests/test_user_registration.py:38]: @felixkiryowa unnecessary change

File: authors/apps/authentication/views.py [hunk 5]
```diff
@@ -18,12 +18,14 @@ class RegistrationAPIView(GenericAPIView):
...
         user = request.data.get('user', {})
+        # Calling the validate user registration data before passing it to the serializer to save it
+
         # The create serializer, validate serializer, save serializer pattern
         # below is common and you will see it a lot throughout this course and
         # your own work later on. Get familiar with it.
         serializer = self.serializer_class(data=user)
-        serializer.is_valid(raise_exception=True)
-        serializer.save()
+        if serializer.is_valid(raise_exception=True):
```
Comment (ja-odur) [inline authors/apps/authentication/views.py:29]: @felixkiryowa please undo the chnages to this file as well

File: authors/apps/profiles/models.py [hunk 6]
```diff
@@ -21,10 +20,10 @@ class Profile(models.Model):
     created_at = models.DateTimeField(auto_now_add=True)
     updated_at = models.DateTimeField(auto_now=True)
 
-
     def __str__(self):
         return self.user
 
+
```
Comment (ja-odur) [inline authors/apps/profiles/models.py:26]: @felixkiryowa necessary changes to this file as well

File: authors/apps/profiles/permissions.py [hunk 7]
```diff
@@ -5,11 +5,11 @@ class IsOwnerOrReadOnly(BasePermission):
     """
     This method allows only the owner who created a profile to view it
     """
-    
+
```
Comment (ja-odur) [inline authors/apps/profiles/permissions.py:8]: @felixkiryowa unnecessary changes here as well

File: authors/apps/profiles/views.py [hunk 8]
```diff
@@ -1,17 +1,14 @@
-
+from django.shortcuts import get_object_or_404
 from rest_framework import status
-from rest_framework.generics import RetrieveAPIView, ListAPIView, CreateAPIView, RetrieveUpdateAPIView
-from rest_framework.permissions import(
-    AllowAny,IsAuthenticatedOrReadOnly, IsAuthenticated
+from rest_framework.generics import ListAPIView, RetrieveUpdateAPIView
+from rest_framework.permissions import (
```
Comment (ja-odur) [inline authors/apps/profiles/views.py:4]: @felixkiryowa remove the parenthesis if importing only one item

**Review by ja-odur — CHANGES_REQUESTED** [review]
Please rebase as well

File: authors/apps/authentication/tests/test_base.py (hunk 1 above)
Comment (felixkiryowa) [reply authors/apps/authentication/tests/test_base.py:12]: I  have fixed  it

File: authors/apps/authentication/serializers.py (hunk 2 above)
Comment (felixkiryowa) [reply authors/apps/authentication/serializers.py]: I have removed it

File: Procfile (hunk 3 above)
Comment (felixkiryowa) [reply Procfile:2]: I have fixed it

File: authors/apps/authentication/views.py (hunk 5 above)
Comment (felixkiryowa) [reply authors/apps/authentication/views.py:29]: I have fixed it

**Review by ezrogha — COMMENTED** [review]
Great work @felixkiryowa, 
Only that I noticed some inconsistency in your postman examples
Example `(400) Username and Password are not provided in the request` validates both the username and password fields returning both error messages
where as
Example `(400)  Invalid password format` validates only the password and returns the password error, yet the username(andrew#@lk) also has issues. 
Shouldn't it return the username as well?

**Review by dannylwe — CHANGES_REQUESTED** [review]
Your build is currently failing on Travis, please fix. Also, kindly rebase.

File: Procfile (hunk 3 above)
Comment (dannylwe) [reply Procfile:2]: @felixkiryowa This line of code still exists and is causing conflicts with develop.

File: authors/apps/authentication/serializers.py [hunk 9]
```diff
@@ -15,11 +16,13 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
         write_only=True
     )
+    email = serializers.EmailField()
+    username = serializers.CharField()
 
     # The client should not be able to send a token along with a registration
     # request. Making `token` read-only handles that for us.
 
-    #making token read-only
+    # making token read-only
```
Comment (dannylwe) [inline authors/apps/authentication/serializers.py:25]: @felixkiryowa this change is unnecessary. Try and limit these types of changes.

File: authors/apps/authentication/tests/test_base.py (hunk 1 above)
Comment (dannylwe) [reply authors/apps/authentication/tests/test_base.py:12]: @felixkiryowa these changes still exist in your codebase. Please fix.

File: authors/apps/authentication/tests/test_base.py [hunk 10]
```diff
@@ -18,17 +20,27 @@ def setUp(self):
...
 
-        if not BaseTestCase.token:
-            self.client.post('/api/users/register/', login_credentials_data, format='json')
-            response1 = self.client.post('/api/users/login/', login_data, format='json')
+        self.token = 'Bearer {}'.format(BaseTestCase.token)
 
-            BaseTestCase.token = response1.data["token"]
-        return BaseTestCase.token 
+
+    # def generate_jwt_token(self):
```
Comment (dannylwe) [inline authors/apps/authentication/tests/test_base.py:33]: @felixkiryowa Can you explain why this code is commented out? Cc: @ja-odur

File: authors/apps/authentication/views.py [hunk 11]
```diff
@@ -18,12 +19,15 @@ class RegistrationAPIView(GenericAPIView):
...
+        # validate_registration(user)
+
         # The create serializer, validate serializer, save serializer pattern
         # below is common and you will see it a lot throughout this course and
         # your own work later on. Get familiar with it.
         serializer = self.serializer_class(data=user)
-        serializer.is_valid(raise_exception=True)
-        serializer.save()
+        if serializer.is_valid(raise_exception=True):
+            serializer.save()
```
Comment (dannylwe) [inline authors/apps/authentication/views.py:30]: @felixkiryowa kindly revert these changes. Remove the if clause above and indention here

**Review by dannylwe — CHANGES_REQUESTED** [review]
